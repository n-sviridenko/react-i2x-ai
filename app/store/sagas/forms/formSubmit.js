import { SubmissionError } from 'redux-form/immutable';
import { take, race, put, call } from 'redux-saga/effects';

export default function* formSubmit({ meta }) {
  const {
    requestAction,
    successActionType,
    failActionType,
    resolve,
    reject,
  } = meta;

  yield put(requestAction);

  const { success } = yield race({
    success: take(successActionType),
    failure: take(failActionType),
  });

  if (success) {
    yield call(resolve, success.payload);
  } else {
    yield call(reject, new SubmissionError());
  }
}
