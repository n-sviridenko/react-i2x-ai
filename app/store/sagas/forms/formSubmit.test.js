/* eslint-disable redux-saga/yield-effects */
import { SubmissionError } from 'redux-form';
import { take, race, call } from 'redux-saga/effects';

import formSubmit from './formSubmit';

describe('formSubmit', () => {
  const requestAction = {
    type: 'TEST_REQUEST',
    payload: { foo: 'bar' },
  };

  const successActionType = 'TEST_SUCCESS';
  const failActionType = 'TEST_FAIL';
  const resolve = () => {};
  const reject = () => {};

  let formSubmitTask;

  beforeEach(() => {
    formSubmitTask = formSubmit({
      meta: {
        requestAction,
        successActionType,
        failActionType,
        resolve,
        reject,
      },
    });

    const requestDescriptor = formSubmitTask.next().value;
    expect(requestDescriptor).toMatchSnapshot();

    const raceDescriptor = formSubmitTask.next().value;
    const expectedRaceDescriptor = race({
      success: take(successActionType),
      failure: take(failActionType),
    });
    expect(raceDescriptor).toEqual(expectedRaceDescriptor);
  });

  it('should execute resolve it successAction was received', () => {
    const actionDescriptor = formSubmitTask.next({ success: { payload: {} } }).value;
    expect(actionDescriptor).toEqual(call(resolve, {}));
  });

  it('should execute resolve it failAction was received', () => {
    const actionDescriptor = formSubmitTask.next({ failure: { payload: {} } }).value;
    expect(actionDescriptor).toEqual(call(reject, new SubmissionError({})));
  });
});
