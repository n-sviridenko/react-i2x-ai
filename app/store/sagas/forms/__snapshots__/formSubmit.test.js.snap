// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formSubmit should execute resolve it failAction was received 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "foo": "bar",
      },
      "type": "TEST_REQUEST",
    },
    "channel": null,
  },
}
`;

exports[`formSubmit should execute resolve it successAction was received 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "foo": "bar",
      },
      "type": "TEST_REQUEST",
    },
    "channel": null,
  },
}
`;
